metadata/ota_metadata/metadata_gen.py

```diff
@@ -11,15 +13,51 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
-
-#!/usr/bin/env python3
-
 import os
-from hashlib import sha256
 import argparse
 import pathlib
+import zstandard
 import igittigitt
+from hashlib import sha256
+
+ZSTD_COMPRESSION_EXTENSION = "zst"
+ZSTD_COMPRESSION_LEVEL = 10
+ZSTD_MULTITHREADS = 2
+CHUNK_SIZE = 4 * (1024**2)  # 4MiB
+
+
+class ZstdSkipFile(Exception):
+    pass
+
+
+def zstd_compress_file(
+    cctx: zstandard.ZstdCompressor,
+    src_fpath: str,
+    dst_fpath: str,
+    *,
+    cmpr_ratio: float,
+    filesize_threshold: int,
+) -> bool:
+    if (src_size := os.path.getsize(src_fpath)) < filesize_threshold:
+        return False  # skip file with too small size
+    try:
+        with open(src_fpath, "rb") as src_f, open(dst_fpath, "wb") as dst_f:
+            with cctx.stream_writer(dst_f, size=src_size) as compressor:
+                while data := src_f.read(CHUNK_SIZE):
+                    compressor.write(data)
+        # drop compressed file if cmpr ratio is too small or compressed failed
+        if (
+            not (compressed_bytes := os.path.getsize(dst_fpath))
+            or src_size / compressed_bytes < cmpr_ratio
+        ):
+            raise ZstdSkipFile
+        return True
+    except Exception:  # cleanup on failure
+        try:
+            os.remove(dst_fpath)
+        except OSError:
+            pass
+        return False
```

Comment:
Only expected exception should be handled. If unintended exception occurred, it should raised.
```suggestion
        with open(src_fpath, "rb") as src_f, open(dst_fpath, "wb") as dst_f:
            with cctx.stream_writer(dst_f, size=src_size) as compressor:
                while data := src_f.read(CHUNK_SIZE):
                    compressor.write(data)
        # drop compressed file if cmpr ratio is too small or compressed failed
        if (
            not (compressed_bytes := os.path.getsize(dst_fpath))
            or src_size / compressed_bytes < cmpr_ratio
        ):
            try:
                os.remove(dst_fpath)
            except OSError:
                pass
            return False
        return True
```